---
title: "Rhythmicity and differential rhythmicity analysis"
author: "Ali Altintas"
date: "2023-02-15"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Rhythmicity_Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width=8,
  fig.height=3, 
  fig.retina = 2
)
```

## Introduction

This package, [cosinoRmixedeffects2](https://github.com/altintasali/cosinoRmixedeffects2), is forked from [cosinoRmixedeffects](https://github.com/maytesuarezfarinas/cosinoRmixedeffects). [cosinoRmixedeffects](https://github.com/maytesuarezfarinas/cosinoRmixedeffects) makes an excellent implementation of mixed effects model to calculate the effect size, confidence intervals, standard errors and p-values using a bootstrapping approach while comparing the rhythmicity parameters *MESOR*, *amplitude* and *acrophase* across two or more groups. In the circadian bioinformatics, this approach is called as **differential rhythmicity** as it compares the **rhythmicity** between 2 or more groups. 

Prior to the differential rhythmicity analysis, a pre-filtering step is required to check whether at least one of the groups show rhythmicity in the measured entity such as gene expression. For example, let's assume that we would like to test the differential rhythmicity of 3 genes (*gene_1*, *gene_2*, *gene_3*) between 2 groups (*group_A*, *group_B*). Below &#10003; and &#10005; shows if a gene is **rhythmic** or **not rhythmic**, respectively.

| Gene          | group_A       | group_B       |
| :-----------: | :-----------: | :-----------: |
| gene_1        | &#10003;      |   &#10003;    |
| gene_2        | &#10003;      |   &#10005;    |
| gene_3        | &#10005;      |   &#10005;    |

Imagine that you ran a differential rhythmicity analysis for *gene_3* and it turned out to be statistically significant. However, remember that *gene_3* did not show any rhytmicity neither in *group_1* nor in *group_2*. This will obviously lead to a false positive discovery. Therefore, genes (or whatever feature you are testing) that do not show any rhythmicity should be discarded from differential rhythmicity analysis. In the toy example above, we should keep *gene_1* and *gene_2* while filtering out *gene_3* for differential rhythmicity analysis.

[cosinoRmixedeffects2](https://github.com/altintasali/cosinoRmixedeffects2) can perform **rhythmicity analysis** while keeping the same methodological principles presented in [cosinoRmixedeffects](https://github.com/maytesuarezfarinas/cosinoRmixedeffects) for **differential rhythmicity analysis**.

### What's new? 

New features

* Rhythmicity analysis
* Improved acrophase calculation
* New parameter `seed`: random seed assignment for reproducible results
* CRAN style R package

Bug fixes (from `cosinoRmixedeffects`)

* Amplitude calculation
* Small improvements and fixes on function parameters

## Installation

Please install `devtools` if you haven't yet.

```{r devtools, eval=FALSE}
install.packages("devtools")
```

Then, you finally install the `cosinoRmixedeffects2` package by using:

```{r install, eval=FALSE}
library(devtools)
install_github("altintasali/cosinoRmixedeffects2", build_vignettes = TRUE, dependencies = TRUE)
```

## Required libraries
```{r libraries}
## this package
library("cosinoRmixedeffects2")

## additional packages for this vignette
if (!require("pacman")) install.packages("pacman")
pacman::p_load("stringr", "ggplot2", "magrittr", 
               "metap", "rstatix", "ggpubr")
```

## Toy dataset

Let's simulate a toy dataset, of which we can test the rhythmicity and differential rhtyhmicity. 

```{r toy_data}
noise_levels <- seq(0.2, 1, by=0.2)
set.seed(2023)

gene1_a <- sapply(noise_levels, function(x){simulate_rhythm(time_series = 1:48,
                                                            mesor = 10, 
                                                            amplitude = 1, 
                                                            acrophase = 6,
                                                            noise = x,
                                                            plot = FALSE)})

gene1_b <- sapply(noise_levels, function(x){simulate_rhythm(time_series = 1:48,
                                                            mesor = 10, 
                                                            amplitude = 2, 
                                                            acrophase = 6,
                                                            noise = x,
                                                            plot = FALSE)})

gene2_a <- sapply(noise_levels, function(x){simulate_rhythm(time_series = 1:48,
                                                            mesor = 10, 
                                                            amplitude = 1, 
                                                            acrophase = 6,
                                                            noise = x,
                                                            plot = FALSE)})

gene2_b <- sapply(noise_levels, function(x){simulate_rhythm(time_series = 1:48,
                                                            mesor = 1, 
                                                            amplitude = 0.000001, 
                                                            acrophase = 6,
                                                            noise = x,
                                                            plot = FALSE)})


gene3_a <- sapply(noise_levels, function(x){simulate_rhythm(time_series = 1:48,
                                                            mesor = 1, 
                                                            amplitude = 0.000001, 
                                                            acrophase = 6,
                                                            noise = x,
                                                            plot = FALSE)})

gene3_b <- sapply(noise_levels, function(x){simulate_rhythm(time_series = 1:48,
                                                            mesor = 1, 
                                                            amplitude = 0.000001, 
                                                            acrophase = 6,
                                                            noise = x,
                                                            plot = FALSE)})

gene1 <- cbind(gene1_a, gene1_b)
rownames(gene1) <- 1:48
colnames(gene1) <- c(paste("a", 1:5, sep = "_"), paste("b", 1:5, sep = "_"))
gene1 <- reshape2::melt(gene1)
gene1 <- data.frame(gene_id = "gene_1", gene1)

gene2 <- cbind(gene2_a, gene2_b)
rownames(gene2) <- 1:48
colnames(gene2) <- c(paste("a", 1:5, sep = "_"), paste("b", 1:5, sep = "_"))
gene2 <- reshape2::melt(gene2)
gene2 <- data.frame(gene_id = "gene_2", gene2)

gene3 <- cbind(gene3_a, gene3_b)
rownames(gene3) <- 1:48
colnames(gene3) <- c(paste("a", 1:5, sep = "_"), paste("b", 1:5, sep = "_"))
gene3 <- reshape2::melt(gene3)
gene3 <- data.frame(gene_id = "gene_3", gene3)

rhy_data <- rbind(gene1, gene2, gene3)

group_ind <- str_split_fixed(rhy_data$Var2, "_", 2)
group_ind <- as.data.frame(group_ind)
rhy_data <- cbind(rhy_data, group_ind)
rhy_data <- rhy_data[, -3]
colnames(rhy_data)  <- c("gene_id", "Time", "value", "group", "individual")
rhy_data <- data.frame(unclass(rhy_data), stringsAsFactors = TRUE)
str(rhy_data)

ggplot(rhy_data, aes(x = Time, y = value)) +
  geom_point(aes(color = group)) +
  facet_grid(~gene_id)

ggplot(rhy_data, aes(x = Time, y = value)) +
  geom_point(aes(color = individual)) +
  facet_grid(~gene_id)

```

## Rhyhmicity analysis

Rhythmicity is defined as the rhythms that has an amplitude bigger than 0. In other words, the null hypothesis for rhythmicity is amplitude being equal to 0. Therefore, we will only check the amplitudes and their associated p-values for genes (*gene_1*, *gene_2*, *gene_3*) within each group (*groupA*, *groupB*).

For faster calculation, calculations are done by running bootstrap 50 times (`nsim = 50`) using 2 threads (`ncpus = 2`). While using real data, I recommend running bootstrapping thousands of times for more precise calculations (e.g. `nsim = 10000`). Setting the maximum available CPU cores will make this process a bit faster (e.g. `ncpus = 8`), however it will still be slow. Be patient and let the computer do the calculations for your (maybe while you are sleeping). 

```{r rhythmicity}
## create input data (group A)
db.model_A <- create.cosinor.param(time = "Time", 
                                   period = 24, 
                                   data = subset(rhy_data, group == "a"))

rhy_data_groupA <- split(db.model_A, f = db.model_A$gene_id)

## create input data (group B)
db.model_B <- create.cosinor.param(time = "Time", 
                                   period = 24, 
                                   data = subset(rhy_data, group == "b"))

rhy_data_groupB <- split(db.model_B, f = db.model_B$gene_id)

## Function: Calculate rhythmicity 
calc_rhy <- function(x){
  fit <- fit.cosinor.mixed(y="value", random="1 | individual", data = x)
  cont_res <- get.contrasts.ci.cosinor(fit = fit, 
                                       contrast.frm = "~rrr + sss", 
                                       nsim = 50, 
                                       ncpus = 2, 
                                       pairwise = FALSE, 
                                       seed = 2023, 
                                       plot = FALSE)
  rhy_res <- data.frame(amp = cont_res["Amplitude","boot.estimate"],
                      relAmp = cont_res["Amplitude","boot.estimate"]/cont_res["MESOR","boot.estimate"],
                      pvalue = cont_res["Amplitude","pvalue"])
  return(rhy_res)
}

## Calculate rhythmicity for each group
rhy_res_groupA <- sapply(rhy_data_groupA, calc_rhy) %>%
  t %>%
  apply(., 2, unlist) %>% 
  as.data.frame
rhy_res_groupA$padj <- p.adjust(rhy_res_groupA$pvalue, method = "BH")
rhy_res_groupA$rhythmic <- rhy_res_groupA$padj < 0.05
rhy_res_groupA

rhy_res_groupB <- sapply(rhy_data_groupB, calc_rhy) %>%
  t %>%
  apply(., 2, unlist) %>% 
  as.data.frame 
rhy_res_groupB$padj <- p.adjust(rhy_res_groupB$pvalue, method = "BH")
rhy_res_groupB$rhythmic <- rhy_res_groupB$padj < 0.05
rhy_res_groupB
```
The table below summarizes the rhytmicity status of each gene within each group.

| Gene          | group_A       | group_B       |
| :-----------: | :-----------: | :-----------: |
| gene_1        | &#10003;      |   &#10003;    |
| gene_2        | &#10003;      |   &#10005;    |
| gene_3        | &#10005;      |   &#10005;    |

## Differential rhythmicity analysis

### Pre-filtering

Before proceeding with the differential rhythmicity analysis, any genes with no rhythmicity should be discarded. There are different ways of filtering out the non-rhythmic genes. Here, I present 2 common approaches: 

#### Boolean filtering

This is the simplest way of filtering. The idea is to include any gene that shows rhythmicity across any of the groups. 

```{r boolean_filter}
keep <- rhy_res_groupA$rhythmic | rhy_res_groupA$rhythmic
rownames(rhy_res_groupA)[keep]
```
#### Combining p-values

Combining p-values is a more statistically robust method. I recommend either Fisher's method or Trippet's method to combine (raw) p-values. Those combined p-values, a.k.a meta p-values, should then be subject to adjustment for multiple testing.

```{r combineP_filter}
pmat <- cbind(rhy_res_groupA$pvalue, rhy_res_groupB$pvalue)

## Meta p-values using Trippet's method
metap <- apply(pmat, 1, function(x){metap::minimump(x)$p})
## Meta p-values using Fisher's method
#metap <- apply(pmat, 1, function(x){metap::sumlog(x)$p})

metap_adj <- p.adjust(metap, method = "BH")
keep <- metap_adj < 0.05
rownames(rhy_res_groupA)[keep]
```

### Differential rhytmicity analysis

For faster calculation, calculations are done by running bootstrap 50 times (`nsim = 50`) using 2 threads (`ncpus = 2`). While using real data, I recommend running bootstrapping thousands of times for more precise calculations (e.g. `nsim = 10000`). Setting the maximum available CPU cores will make this process a bit faster (e.g. `ncpus = 8`), however it will still be slow. Be patient and let the computer do the calculations for your (maybe while you are sleeping). 

```{r diff_rhy}
## Prepare data
genes2keep <- rownames(rhy_res_groupA)[keep]

rhy_data_filt <- rhy_data[rhy_data$gene_id %in% genes2keep, ]

db.model <- create.cosinor.param(time = "Time", 
                                   period = 24, 
                                   data = rhy_data_filt)

db.model_split <- split(db.model, f = droplevels(db.model$gene_id))

## Function: Calculate differential rhythmicity 
calc_diff_rhy <- function(x){
  message(paste0("Differential rhytmicty: ", x))
  
  X <- db.model_split[[x]]
  fit <- fit.cosinor.mixed(x = "group", 
                           y = "value", 
                           random="1 | individual",
                           data = X)
  
  db.means <- get.means.ci.cosinor(fit = fit, 
                                   contrast.frm = "~group", 
                                   nsim = 50, 
                                   ncpus = 2, 
                                   pairwise = FALSE, 
                                   seed = 2023)
  
  db.delta <- get.contrasts.ci.cosinor(fit = fit, 
                                       contrast.frm = "~group",
                                       nsim = 50,
                                       ncpus = 2, 
                                       pairwise = TRUE, 
                                       seed = 2023, 
                                       plot = FALSE)
  diff_rhy_res <- data.frame(gene = x,
                             MESOR = db.delta[1, "boot.estimate"], 
                             MESOR_p = db.delta[1, "pvalue"], 
                             Amplitude = db.delta[2, "boot.estimate"], 
                             Amplitude_p = db.delta[2, "pvalue"], 
                             Acrophase = db.delta[3, "boot.estimate"], 
                             Acrophase_p = db.delta[3, "pvalue"])
  
  out <- list(db.means = db.means, db.delta = db.delta, table = diff_rhy_res)
  return(out)
}

## Function: Calculate differential rhythmicity 
diff_rhy <- lapply(genes2keep, calc_diff_rhy)
names(diff_rhy) <- genes2keep

diff_rhy_table <- do.call(rbind, lapply(diff_rhy, function(x){x$table}))
head(diff_rhy_table)
```

You can also plot your results for the gene of your choice by:

```{r plot_res}
gene2plot <- "gene_1"

p <- ggplot(diff_rhy[[gene2plot]]$db.means, aes(x = group, y = MEAN, color = group)) +
  geom_point(position = position_dodge(width=1)) +
  scale_color_manual(values = c(
    a="forestgreen", 
    b="orange"))+
  geom_errorbar(aes(ymin=`2.5 %`, ymax=`97.5 %`), position = position_dodge(width = 1))+
  theme_bw()+
  facet_wrap(~Param, scales = "free_y") +
  guides(color=guide_legend(title="group"))+
  labs(x="group", y="Mean (CI 95%)") + 
  theme(strip.background = element_rect(color="black", fill="white", linetype=0))+
  theme(strip.text.x = element_text(size = 14))
p

stat.test <- stat.test.stars(db.delta = diff_rhy[[gene2plot]]$db.delta, 
                             db.means = diff_rhy[[gene2plot]]$db.means, 
                             contrast.frm = "~group")

p <- p + ggpubr::stat_pvalue_manual(stat.test, hide.ns = T, step.increase = 0.05, tip.length = 0.03)

p

```

